[PEP 741] gh-107954: Add PyInitConfig C API (#110176)
python/cpython: Doc/c-api/config.rst

.. highlight:: c

.. _config-c-api:

********************
Python Configuration
********************

.. versionadded:: 3.13

API part of the limited C API version 3.13 to configure the Python
initialization and get the Python runtime configuration.

See also :ref:`Python Initialization Configuration <init-config>`.

Initialize Python
=================

.. c:struct:: PyInitConfig

   Opaque structure to configure the Python initialization.

.. c:function:: PyInitConfig* PyInitConfig_Python_New(void)

   Create a new initialization configuration using :ref:`Python Configuration
   <init-python-config>` default values.

   It must be freed by :c:func:`PyInitConfig_Free`.

   Return ``NULL`` on memory allocation failure.

.. c:function:: PyInitConfig* PyInitConfig_Isolated_New(void)

   Similar to :c:func:`PyInitConfig_Python_New`, but use :ref:`Isolated
   Configuration <init-isolated-conf>` default values.

.. c:function:: void PyInitConfig_Free(PyInitConfig *config)

   Free memory of a initialization configuration.

> .. c:function:: int PyInitConfig_SetInt(PyInitConfig *config, const char *key, int64_t value)

Reviewer: It's not clear that `key` should be a name of a `PyConfig` member. (Should it?)

   Set an integer configuration option.

   Return 0 on success, or return -1 on error.

.. c:function:: int PyInitConfig_SetStr(PyInitConfig *config, const char *key, const char *value)

>    Set a string configuration option from a bytes string.

Reviewer: I'd prefer to be explicit in new docs:
```suggestion
   Set a string configuration option from a NUL-terminated bytes string.
```

   The bytes string is decoded by Py_DecodeLocale(). Preinitialize Python if
   needed to ensure that encodings are properly configured.

>    Return 0 on success, or return -1 on error.

Reviewer: And I'd make it more obvious that that the error is stored in the `PyInitConfig`:
```suggestion
   Return 0 on success.
   On error, return -1. The error can be retrieved with
   :c:expr:`PyInitConfig_GetErrorMsg(config)`
```

Author: I updated the doc.

.. c:function:: int PyInitConfig_SetWStr(PyInitConfig *config, const char *key, const wchar_t *value)

   Set a string configuration option from a wide string.

>    Preinitialize Python if needed.

Reviewer: It's not clear if the function preinitializes Python, or if the *user* needs to do it if needed.
```suggestion
   If Python is not yet preinitialized, this function preinitializes it.
```

   Return ``0`` on success, or return ``-1`` on error.

.. c:function:: int PyInitConfig_SetStrList(PyInitConfig *config, const char *key, size_t length, char * const *items)

>    Set a string list configuration option from bytes strings.

Reviewer: ```suggestion
   Set a string list configuration option from an array of NUL-terminated bytes strings.
```

   The bytes strings are decoded by Py_DecodeLocale(). Preinitialize Python if
   needed to ensure that encodings are properly configured.

   Return ``0`` on success, or return ``-1`` on error.

.. c:function:: int PyInitConfig_SetWStrList(PyInitConfig *config, const char *key, size_t length, wchar_t * const *items)

   Set a string list configuration option from a wide strings.
   Preinitialize Python if needed.
   Return 0 on success, or return -1 on error.

.. c:function:: int Py_InitializeFromInitConfig(PyInitConfig *config)

   Initialize Python from the initialization configuration.

   Return ``0`` on success.

   Return ``-1`` if Python wants to exit and on error. Call

>    :c:func:`Py_ExitWithInitConfig` in this case.

Reviewer: What happens if you don't call it? (If it needs to be called, why does `Py_InitializeFromInitConfig` not call it itself?)

Author: The idea is to let the caller decide how to handle an error or an exit code. When you embed Python, you may not want the process to exit. It's a similar design than `Py_InitializeFromConfig()` and the `PyStatus` C API.

> What happens if you don't call it?

Python is in an undefined state. The error/exit code stays in PyInitConfig. It's safer to handle the init error :-)
